javascript/packages/fury/lib/gen/map.ts

```diff
@@ -21,77 +21,237 @@ import { MapTypeDescription, TypeDescription } from "../description";
+    if (header !== 0) {
+      if (header & MapFlags.HAS_NULL) {
+        if (v === null || v === undefined) {
+          this.fury.binaryWriter.uint8(RefFlags.NullFlag);
+        }
+      }
+      if (header & MapFlags.TRACKING_REF) {
+        const keyRef = this.fury.referenceResolver.existsWriteObject(v);
+        if (keyRef !== undefined) {
+          this.fury.binaryWriter.uint8(RefFlags.RefFlag);
+          this.fury.binaryWriter.uint16(keyRef);
+        } else {
+          this.fury.binaryWriter.uint8(RefFlags.RefValueFlag);
+        }
+      } else {
+        this.fury.binaryWriter.uint8(RefFlags.NotNullValueFlag);
+      }
+    }
+  }
+
+  write(value: Map<any, any>) {
+    const mapChunkWriter = new MapChunkWriter(this.fury);
+    this.fury.binaryWriter.varInt32(value.size);
+    for (const [k, v] of value.entries()) {
+      const keySerializer = this.keySerializer !== null ? this.keySerializer : this.fury.classResolver.getSerializerByData(k);
+      const valueSerializer = this.valueSerializer !== null ? this.valueSerializer : this.fury.classResolver.getSerializerByData(v);
 
-    const result = this.scope.uniqueName("result");
-    const idx = this.scope.uniqueName("idx");
-    const len = this.scope.uniqueName("len");
+      const header = mapChunkWriter.next(
+        mapChunkWriter.makeFlag(keySerializer!.meta.typeId!, k == null ? 1 : 0, keySerializer!.meta.needToWriteRef ? 1 : 0),
+        mapChunkWriter.makeFlag(valueSerializer!.meta.typeId!, v == null ? 1 : 0, valueSerializer!.meta.needToWriteRef ? 1 : 0)
+      );
 
+      this.writeHead(header >> 4, k);
+      keySerializer!.writeInner(k);
+      this.writeHead(header & 0b00001111, v);
+      valueSerializer!.writeInner(v);
+    }
+    mapChunkWriter.endChunk();
+  }
+
+  private readElement(header: number, serializer: Serializer | null) {
+    if (header === 0) {
+      return serializer!.readInner(false);
+    }
+    const isSame = !(header & MapFlags.NOT_SAME_TYPE);
+    const includeNone = header & MapFlags.HAS_NULL;
+    const trackingRef = header & MapFlags.TRACKING_REF;
+
+    let flag = 0;
+    if (trackingRef || includeNone) {
+      flag = this.fury.binaryReader.uint8();
+    }
+    if (!isSame) {
+      serializer = this.fury.classResolver.getSerializerByType(this.fury.binaryReader.int16());
+    }
+    switch (flag) {
+      case RefFlags.RefValueFlag:
+        return serializer!.readInner(true);
+      case RefFlags.RefFlag:
+        return this.fury.referenceResolver.getReadObject(this.fury.binaryReader.varUInt32());
+      case RefFlags.NullFlag:
+        return null;
+      case RefFlags.NotNullValueFlag:
+        return serializer!.readInner(false);
+    }
+  }
+
+  read(fromRef: boolean): any {
+    let count = this.fury.binaryReader.varInt32();
+    const result = new Map();
+    if (fromRef) {
+      this.fury.referenceResolver.reference(result);
+    }
+    while (count > 0) {
+      const header = this.fury.binaryReader.uint16();
```

Comment:
Do we need to rename it as `read/writeXXX` in future?

Reply:
I think it is uncessary, unlike java the binaryWriter and binaryReader  are separate classes.  In `v8` , if the count of a object's property greater than 30, the object will degenerate into hashtable, which is very very shlow. So I  split the class `MemoryBuffer` into two  separate class.